protocol.md

```diff
@@ -726,6 +727,120 @@ Upload-Length: 11
 Upload-Concat: final;/files/a /files/b
 ```
 
+### Client Tag
+
+With this extension, Clients can provide a Tag for an upload - with which they
+can later retrieve the URL for the created resource.
+
+This allows Clients to resume an upload initiated through the [Creation With Upload](#creation-with-upload)
+or [Creation](#creation) extensions for which they did not receive the response with
+the newly created resource's URL (for example due to a broken connection).
+
+If the Server supports this extension, it MUST add `client-tag` to the `Tus-Extension`
+header.
+
+To initiate an upload with Client Tag, Clients MUST include an `Upload-Tag`
+header with the initial `POST` request (as specified in the [Creation](#creation) and
+[Creation With Upload](#creation-with-upload) extensions) made to the Upload URL.
+
+Clients can then send a `HEAD` request with the same `Upload-Tag` header to the
+Upload URL, which - upon success - MUST respond with the `200 OK` or `204 No Content`
+status, and the resource's URL in the response's `Location` header.
+
+A successful response MUST also contain the `Tus-Version` header. An `Upload-Offset` header
+that indicates how many bytes have already been received by the server MAY also be returned.
+
+Clients can then use the returned `Location` to resume the upload.
+
+If an `Upload-Tag` is unknown or no longer known by the Server, it MUST
+respond to `HEAD` requests with the `404 Not Found` or `410 Gone` status.
+
+If an `Upload-Tag` is already in use for a different, ongoing upload, the Server
+MUST respond to the `POST` request initiating the upload with a `409 Conflict` status.
+
+#### Security Considerations
+
+To protect against attackers trying to guess an `Upload-Tag` and subsequently
+use it to inject malicious content into an upload, Servers MUST take measures to
+ensure that only the party that created a resource with an `Upload-Tag` can use it
+for subseqeuent `PATCH` and `HEAD` requests.
+
+If uploading to a Server is only possible for authenticated users, the Server can
+satisfy this requirement by leveraging the available authentication information to
+bind the `Upload-Tag` to a particular user, so that only that user can use it.
+
+Servers MUST return a `403 Forbidden` status for unauthenticated requests that
+contain an `Upload-Tag` header.
+
+#### Headers
+
+##### Upload-Tag
+
+The `Upload-Tag` request header MUST be an identifier created by the
+Client that's unique within its storage space, to avoid collisions.
```

Comment:
> unique within its storage space

I think this wording is a bit weird. As an outside user, who has not been involved into this discussion, I would not understand what uniqueness in its storage space means. However, I am not sure what better wording would look like.

Reply:
Agreed. Could we change it to something like the following?

> The `Upload-Tag` request header MUST be an identifier created by the Client that's unique to avoid collisions. If a collision is detected the Server MUST respond with a `409 Conflict` status

The reason the wording was in there to begin with was probably my fault as tusdotnet has the concept of storage spaces where each user/domain/whatever can store files differently.

Reply:
Good proposal. Maybe we should be even more simple:

> The Upload-Tag request header MUST be an identifier created by the Client. The identifier SHOULD be unique to avoid collisions with identifiers created by other Clients.

I use SHOULD instead of MUST here since a client cannot be entirely sure that an identifier is not used by a different client. Does that make sense?

Reply:
It absolutely does. Good catch. 👍

Reply:
> The Upload-Tag request header MUST be an identifier created by the Client. The identifier SHOULD be unique to avoid collisions with identifiers created by other Clients.

I'm going to blatantly steal these two sentences and put them in the next commit. 👍